lib/networkx/multidigraph.rb

```diff
@@ -0,0 +1,233 @@
+module NetworkX
+  class MultiDiGraph < DiGraph
+    # Returns a new key
+    #
+    # @param node_1 [Object] the first node of a given edge
+    # @param node_2 [Object] the second node of a given edge
+    def new_edge_key(node_1, node_2)
+      return 0 if @adj[node_1][node_2].nil?
+      key = @adj[node_1][node_2].length
+      key += 1 while @adj[node_1][node_2].key?(key)
+      key
+    end
+
+    # Adds the respective edge
+    #
+    # @example Add an edge with attribute name
+    #   graph.add_edge(node1, node2, name: "Edge1")
+    #
+    # @example Add an edge with no attribute
+    #   graph.add_edge("Bangalore", "Chennai")
+    #
+    # @param node_1 [Object] the first node of the edge
+    # @param node_2 [Object] the second node of the edge
+    # @param edge_attrs [Hash{ Object => Object }] the hash of the edge attributes
+    def add_edge(node_1, node_2, **edge_attrs)
+      add_node(node_1)
+      add_node(node_2)
+      key = new_edge_key(node_1, node_2)
+      all_edge_attrs = @adj[node_1][node_2] || {}
+      all_edge_attrs[key] = edge_attrs
+      @adj[node_1][node_2] = all_edge_attrs
+      @pred[node_2][node_1] = all_edge_attrs
+    end
+
+    # Removes edge from the graph
+    #
+    # @example
+    #   graph.remove_edge('Noida', 'Bangalore')
+    #
+    # @param node_1 [Object] the first node of the edge
+    # @param node_2 [Object] the second node of the edge
+    def remove_edge(node_1, node_2, key=nil)
+      if key.nil?
+        super(node_1, node_2)
+        return
+      end
+      raise KeyError, "#{node_1} is not a valid node." unless @nodes.key?(node_1)
+      raise KeyError, "#{node_2} is not a valid node" unless @nodes.key?(node_2)
+      raise KeyError, 'The given edge is not a valid one.' unless @adj[node_1].key?(node_2)
+      raise KeyError, 'The given edge is not a valid one.' unless @adj[node_1][node_2].key?(key)
+      @adj[node_1][node_2].delete(key)
+      @pred[node_2][node_1].delete(key)
+    end
+
+    # Checks if the the edge consisting of two nodes is present in the graph
+    #
+    # @example
+    #   graph.edge?(node_1, node_2)
+    #
+    # @param node_1 [Object] the first node of the edge to be checked
+    # @param node_2 [Object] the second node of the edge to be checked
+    # @param key [Integer] the key of the given edge
+    def edge?(node_1, node_2, key=nil)
+      super(node_1, node_2) if key.nil?
+      return true if @nodes.key?(node_1) && @adj[node_1].key?(node_2) && @adj[node_1][node_2].key?(key)
+      false
+    end
+
+    # Returns the undirected version of the graph
+    #
+    # @example
+    #   graph.to_undirected
+    def to_undirected
+      graph = NetworkX::Graph.new(@graph)
+      @nodes.each { |node, node_attr| graph.add_node(node, node_attr) }
```

Comment:
Isn't `MultiDiGraph#to_undirected` supposed to written ` `Multigraph`? Have a good look at this.

Reply:
Hmm, had a look at the `to_multigraph` method. But in networkx, the `#to_undirected` is intended to return `MultiGraph`.

Reference: https://networkx.github.io/documentation/stable/reference/classes/generated/networkx.MultiDiGraph.to_undirected.html

@MridulS @v0dro - Any comments on which one might be more intuitive?